art/defences/detector/poison/activation_defence.py

```diff
@@ -331,6 +350,86 @@ def analyze_clusters(self, **kwargs) -> Tuple[Dict[str, Any], np.ndarray]:
 
         return report, self.assigned_clean_by_class
 
+    def exclusionary_reclassification(self, report):
+        """
+        This function perform exclusionary reclassification. Based on the ex_re_threshold,
+        suspicious clusters will be rechecked. If they remain suspicious, the suspected source
+        class will be added to the report and the data will be relabelled. The new labels are stored
+        in self.y_train_relabelled
+
+        :param report: A dictionary containing defence params as well as the class clusters and their suspiciousness.
+        :return: report where the report is a dict object
+        """
+        self.y_train_relabelled = np.copy(self.y_train)  # Copy the data to avoid overwriting user objects
+        # used for relabeling the data
+        is_onehot = False
+        if len(np.shape(self.y_train)) == 2:
+            is_onehot = True
+
+        logger.info("Performing Exclusionary Reclassification with a threshold of %s", self.ex_re_threshold)
+        logger.info("Data will be relabelled internally. Access the y_train_relabelled attribute to get new labels")
+        # Train a new classifier with the unsuspicious clusters
+        cloned_classifier = (
+            self.classifier.clone_for_refitting()
+        )  # Get a classifier with the same training setup, but new weights
+        filtered_x = self.x_train[np.array(self.is_clean_lst) == 1]
+        filtered_y = self.y_train[np.array(self.is_clean_lst) == 1]
+
+        if len(filtered_x) == 0:
+            logger.warning("All of the data is marked as suspicious. Unable to perform exclusionary reclassification")
+            return report
+
+        cloned_classifier.fit(filtered_x, filtered_y)
+
+        # Test on the suspicious clusters
+        n_train = len(self.x_train)
+        indices_by_class = self._segment_by_class(np.arange(n_train), self.y_train)
+        indicies_by_cluster: List[List[np.ndarray]] = [  # type: ignore
+            [[] for _ in range(self.nb_clusters)] for _ in range(self.classifier.nb_classes)  # type: ignore
```

Comment:
```suggestion
        indicies_by_cluster: List[List[List]] = [
            [[] for _ in range(self.nb_clusters)] for _ in range(self.classifier.nb_classes)
```
```suggestion
        indicies_by_cluster: List[List[np.ndarray]] = [  # type: ignore
            [[] for _ in range(self.nb_clusters)] for _ in range(self.classifier.nb_classes)  # type: ignore
```

Reply:
Do you want it as a list because its more general?

Reply:
Oh, I remember that I was basing my code off 

```
x_raw_by_class = self._segment_by_class(x_raw, self.y_train)
x_raw_by_cluster: List[List[np.ndarray]] = [  # type: ignore
          [[] for _ in range(self.nb_clusters)] for _ in range(self.classifier.nb_classes)  # type: ignore
      ]
```
with a minor modification to get the indices instead

Reply:
Yes, it seemed necessary to have the return type correctly defined.